### ISSUE TITLE
Add support for CodeStar Connections

### ISSUE THREAD

**Issue by jrobison-sb** [hashicorp/terraform-provider-aws#15453]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS [recommends](https://docs.aws.amazon.com/codepipeline/latest/userguide/update-github-action-connections.html) that CodePipeline connections to third party services like GitHub and BitBucket should be done using CodeStar Connections. It would be nice if Terraform supported this.

### New Resources and Data Sources

- [x] resource: aws_codestarconnections_connection (#15990)
- [ ] tag support for codestarconnections service
- [x] resource: aws_codestarconnections_host
- [ ] data source: aws_codestarconnections_connection
- [ ] data source: aws_codestarconnections_host

### References

* https://docs.aws.amazon.com/codestar-connections/latest/APIReference/Welcome.html

**Comment by bflad** [comment]
Related:

* [Codepipeline Bitbucket Integration](https://github.com/terraform-providers/terraform-provider-aws/issues/11389)
* [#11389 Add codestar connection resource](https://github.com/terraform-providers/terraform-provider-aws/pull/12637)

Please note that in this case we'll expect these resources to be named `aws_codestarconnections_XXX` to follow the conventions in the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/contribution-checklists.md#new-resource).

**Comment by mwarkentin** [comment]
> The GitHub version 1 action is not recommended and will soon no longer be supported.

Sounds like this will be the only supported connection type soon.

**Comment by mwarkentin** [comment]
Might be nice to have a `codestar_connection` data source as well so that you could look them up by name and reference their ARN in CodePipeline actions.

**Comment by shuheiktgw** [comment]
I've started adding CodeStar Connections to TF so please upvote the PR below for the faster review! 🙏 

https://github.com/terraform-providers/terraform-provider-aws/pull/15960

**Comment by shuheiktgw** [comment]
Added the second part of the PR: https://github.com/hashicorp/terraform-provider-aws/pull/15990

I'm going to work on `aws_codestarconnections_host` once the PR above is merged!

**Comment by shuheiktgw** [comment]
Added tag support for CodeStar connection! 🎉 
https://github.com/hashicorp/terraform-provider-aws/pull/16835

**Comment by shuheiktgw** [comment]
Added support for CodeStar host!
https://github.com/hashicorp/terraform-provider-aws/pull/16918

**Comment by onitake** [comment]
> Might be nice to have a `codestar_connection` data source as well so that you could look them up by name and reference their ARN in CodePipeline actions.

@shuheiktgw Will you be adding a `aws_codestarconnections_connection` data source as well?
If not, what would be the recommended way to obtain the ARN in a separate Terraform script? Add the `aws_codestarconnections_connection` resource to every script, or copy the ARN?

It doesn't look like the ARN has a predictable value, so it can't be constructed in-place.

**Comment by shuheiktgw** [comment]
@onitake Sure, I'll add `aws_codestarconnections_connection` data 👍

**Comment by DrFaust92** [comment]
Closing this as most (if not all were resource were covered) if a more specific resource/feature is needed please create a dedicated ticket
